modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,709 @@
+proc inner(A: [?Adom], B: [?Bdom]) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+
+  return + reduce(A[..]*B[..]);
+}
+
+
+/* Outer product of 2 vectors */
+proc outer(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+
+  var C: [{Adom.dim(1), Bdom.dim(1)}] eltType;
+  forall (i,j) in C.domain do
+    C[i, j] = A[i]*B[j];
+  return C;
+}
+
+
+pragma "no doc"
+/* Generic matrix-vector multiplication */
+proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType)
+  where !isBLASType(eltType)
+{
+  if Adom.rank != 2 || Xdom.rank != 1 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var C: [Xdom] eltType;
+
+  // naive algorithm
+  forall i in Xdom do
+    C[i] = + reduce (A[i,..]*X[..]);
+
+  return C;
+}
+
+pragma "no doc"
+/* Generic matrix-matrix multiplication */
+proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
+  where !isBLASType(eltType)
+{
+  if Adom.rank != 2 || Bdom.rank != 2 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var C: [Adom.dim(1), Bdom.dim(2)] eltType;
+
+  // naive algorithm
+  forall (i,j) in C.domain do
+    C[i,j] = + reduce (A[i,..]*B[..,j]);
+
+  return C;
+}
+
+
+/* Return the matrix ``A`` to the ``bth`` power, where ``b`` is a positive
+   integral type. */
+proc matPow(A: [], b) where isNumeric(b) {
+  if !isIntegral(b) then
+    // TODO -- support all reals with Sylvester's formula
+    compilerError("matPow only support powers of integers");
+  if !isSquare(A) then
+    halt("Array not square");
+
+  return _expBySquaring(A, b);
+}
+
+
+pragma "no doc"
+/* Exponentiate by squaring recursively */
+private proc _expBySquaring(x: ?t, n): t {
+    // TODO -- _expBySquaring(pinv(x), -n);
+    if n < 0  then halt("Negative powers not yet supported");
+    else if n == 0  then return eye(x.domain);
+    else if n == 1  then return x;
+    else if n%2 == 0  then return _expBySquaring(dot(x, x), n / 2);
+    else return dot(x, _expBySquaring(dot(x, x), (n - 1) / 2));
+}
+
+
+/* Return element-wise power: ``a**B`` */
+proc matPow(a, B: [?Bdom] ?eltType) where isNumeric(a) {
+  var C: [Bdom] eltType = a**B;
+  return C;
+}
+
+
+/* Return element-wise exponential: ``e**A[i,j]`` */
+proc matExp(A: [?Dom] ?eltType) {
+  var B: [Dom] eltType = e**A;
+  return B;
+}
+
+
+pragma "no doc"
+/* Integer overload */
+proc matExp(A: [?Dom] ?eltType) where isIntegralType(eltType) {
+  var B: [Dom] real = e**A;
+  return B;
+}
+
+
+/* Return cross-product of vectors ``A`` and ``B`` with domains of size ``3`` */
+proc cross(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+  if Adom.size != 3 || Bdom.size != 3 then
+    halt("cross() expects arrays of 3 elements");
+
+  var C = Vector(Adom.size, eltType);
+
+  C[0] = A[1]*B[2] - A[2]*B[1];
+  C[1] = A[2]*B[0] - A[0]*B[2];
+  C[2] = A[0]*B[1] - A[1]*B[0];
```

Comment:
I don't think we should assume ``A`` and ``B`` are 0-based or unstrided. You may want to use ``orderToIndex`` or something here.